vayesta/ewf/urdm.py

```diff
@@ -208,11 +211,494 @@ def make_rdm1_ccsd_global_wf(emb, ao_basis=False, with_mf=True, t_as_lambda=Fals
+    # Make projected WF available via remote memory access
+    if mpi:
+        # TODO: use L-amplitudes of cluster X and T-amplitudes,
+        # Only send T-amplitudes via RMA?
+        rma = {x.id: x.results.pwf.pack() for x in emb.get_fragments(active=True, mpi_rank=mpi.rank)}
+        rma = mpi.create_rma_dict(rma)
+
+    # --- Loop over pairs of fragments and add projected density-matrix contributions:
+    dooa = np.zeros((nocca, nocca))
+    doob = np.zeros((noccb, noccb))
+    dvva = np.zeros((nvira, nvira))
+    dvvb = np.zeros((nvirb, nvirb))
+
+    if with_t1:
+        dvoa = np.zeros((nvira, nocca))
+        dvob = np.zeros((nvirb, noccb))
+
+    xfilter = dict(sym_parent=None) if use_sym else {}
+    for x in emb.get_fragments(active=True, mpi_rank=mpi.rank, **xfilter):
+        wfx = x.results.pwf.as_ccsd()
+        if not late_t2_sym:
+            wfx = wfx.restore()
+
+        t2aa = wfx.t2aa
+        t2ab = wfx.t2ab
+        t2bb = wfx.t2bb
+        if late_t2_sym:
+            t2ba = wfx.t2ba
+
+        # Intermediates: leave left index in cluster-x basis:
+        dooxa = np.zeros((x.cluster.nocc_active[0], emb.nocc[0]))
+        dooxb = np.zeros((x.cluster.nocc_active[1], emb.nocc[1]))
+        dvvxa = np.zeros((x.cluster.nvir_active[0], emb.nvir[0]))
+        dvvxb = np.zeros((x.cluster.nvir_active[1], emb.nvir[1]))
+        dvoxa = np.zeros((x.cluster.nvir_active[0], x.cluster.nocc_active[0]))
+        dvoxb = np.zeros((x.cluster.nvir_active[1], x.cluster.nocc_active[1]))
+
+        cx_occ_a, cx_occ_b = x.get_overlap('mo[occ]|cluster[occ]')
+        cx_vir_a, cx_vir_b = x.get_overlap('mo[vir]|cluster[vir]')
+
+        # Loop over ALL fragments y:
+        for y in emb.get_fragments(active=True):
+            if mpi:
+                if y.solver == 'MP2':
+                    wfy = UMP2_WaveFunction.unpack(rma[y.id]).as_ccsd()
+                else:
+                    wfy = UCCSD_WaveFunction.unpack(rma[y.id])
+            else:
+                wfy = y.results.pwf.as_ccsd()
+            if not late_t2_sym:
+                wfy = wfy.restore()
+
+            # Constructing these overlap matrices scales as N(AO)^2,
+            # however they are cashed and will only be calculated N(frag) times
+            cy_occ_a, cy_occ_b = y.get_overlap('mo[occ]|cluster[occ]')
+            cy_vir_a, cy_vir_b = y.get_overlap('mo[vir]|cluster[vir]')
+            # Overlap between cluster-x and cluster-y:
+            rxy_occ_a, rxy_occ_b = np.dot(cx_occ_a.T, cy_occ_a), np.dot(cx_occ_b.T, cy_occ_b)
+            rxy_vir_a, rxy_vir_b = np.dot(cx_vir_a.T, cy_vir_a), np.dot(cx_vir_b.T, cy_vir_b)
+
+            if svd_tol is not None:
+                def svd(a,spin):
+                    assert(spin == 'a' or spin == 'b')
+                    if spin == 'a':
+                        nonlocal total_sv_a, kept_sv_a
+                    elif spin == 'b':
+                        nonlocal total_sv_b, kept_sv_b
```

Comment:
No need to differentiate between `total_sv_a` and `total_sv_b` (same for `kept_sv`) - just add them alltogether and get rid of the spin dependence of `svd`

Reply:
Somewhere around ~400 orbitals.

![image](https://user-images.githubusercontent.com/89935194/181313669-0472512e-44f9-4290-bdd4-88ce1e48ee23.png)

Reply:
The optimised routine is becomes faster somewhere around ~400 orbitals.

![image](https://user-images.githubusercontent.com/89935194/181313669-0472512e-44f9-4290-bdd4-88ce1e48ee23.png)